File: python/open_guardrail/guards/attorney_client_privilege.py
"""Detect and protect privileged communication markers."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_PRIVILEGE = [
    re.compile(r"\battorney[- ]client\s+privilege\b", re.I),
    re.compile(r"\bprivileged\s+and\s+confidential\b", re.I),
    re.compile(r"\bwork\s+product\b", re.I),
    re.compile(r"\blitigation\s+hold\b", re.I),
    re.compile(r"\blegal\s+hold\b", re.I),
    re.compile(r"\bunder\s+privilege\b", re.I),
    re.compile(r"\bprivileged\s+communication\b", re.I),
    re.compile(r"\bprotected\s+by\s+attorney\b", re.I),
]

_SHARING = [
    re.compile(r"\bforward\s+this\b", re.I),
    re.compile(r"\bshare\s+with\b", re.I),
    re.compile(r"\bCC:", re.I),
    re.compile(r"\bsend\s+to\b", re.I),
    re.compile(r"\bpost\s+this\b", re.I),
    re.compile(r"\bmake\s+public\b", re.I),
]


class _AttorneyClientPrivilege:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "attorney-client-privilege"
        self.action = action

    def check(self, text: str, stage: str = "output") -> GuardResult:
        start = time.perf_counter()
        priv: list[str] = []
        for p in _PRIVILEGE:
            m = p.search(text)
            if m:
                priv.append(m.group())
        sharing: list[str] = []
        for p in _SHARING:
            m = p.search(text)
            if m:
                sharing.append(m.group())
        triggered = len(priv) > 0 and len(sharing) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="attorney-client-privilege",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=f'Privileged content may be shared: "{priv[0]}" + "{sharing[0]}"' if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"privilege_markers": priv, "sharing_indicators": sharing} if triggered else None,
        )


def attorney_client_privilege(*, action: str = "block") -> _AttorneyClientPrivilege:
    return _AttorneyClientPrivilege(action=action)

File: packages/guards/src/attorney-client-privilege.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

interface AttorneyClientPrivilegeOptions {
  action: 'block' | 'warn';
}

const PRIVILEGE_PATTERNS: RegExp[] = [
  /\battorney[- ]client\s+privilege\b/i,
  /\bprivileged\s+and\s+confidential\b/i,
  /\bwork\s+product\b/i,
  /\blitigation\s+hold\b/i,
  /\blegal\s+hold\b/i,
  /\bunder\s+privilege\b/i,
  /\bprivileged\s+communication\b/i,
  /\bprotected\s+by\s+attorney\b/i,
];

const SHARING_PATTERNS: RegExp[] = [
  /\bforward\s+this\b/i,
  /\bshare\s+with\b/i,
  /\bCC:/i,
  /\bsend\s+to\b/i,
  /\bpost\s+this\b/i,
  /\bmake\s+public\b/i,
];

export function attorneyClientPrivilege(options: AttorneyClientPrivilegeOptions): Guard {
  return {
    name: 'attorney-client-privilege',
    version: '0.1.0',
    description: 'Detect and protect privileged communication markers',
    category: 'privacy',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const privilegeFound: string[] = [];
      for (const p of PRIVILEGE_PATTERNS) {
        const m = p.exec(text);
        if (m) privilegeFound.push(m[0]);
      }
      const sharingFound: string[] = [];
      for (const p of SHARING_PATTERNS) {
        const m = p.exec(text);
        if (m) sharingFound.push(m[0]);
      }
      const triggered = privilegeFound.length > 0 && sharingFound.length > 0;
      return {
        guardName: 'attorney-client-privilege',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        message: triggered
          ? `Privileged content may be shared: "${privilegeFound[0]}" + "${sharingFound[0]}"`
          : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered
          ? { privilegeMarkers: privilegeFound, sharingIndicators: sharingFound }
          : undefined,
      };
    },
  };
}
